service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/cloudcontext/DeleteResourcesStep.java

```diff
@@ -0,0 +1,143 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DeleteResourcesStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(DeleteResourcesStep.class);
+  private final CloudContextService cloudContextService;
+  private final ControlledResourceService controlledResourceService;
+  private final AuthenticatedUserRequest userRequest;
+  private final UUID workspaceUuid;
+
+  public DeleteResourcesStep(
+      CloudContextService cloudContextService,
+      ControlledResourceService controlledResourceService,
+      AuthenticatedUserRequest userRequest,
+      UUID workspaceUuid) {
+    this.cloudContextService = cloudContextService;
+    this.controlledResourceService = controlledResourceService;
+    this.userRequest = userRequest;
+    this.workspaceUuid = workspaceUuid;
+  }
+
+  @Override
+  //  @SuppressWarnings("unchecked")
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    // Jackson does not deserialize the list properly with a TypeReference. The unchecked for works.
+    ResourceDeleteFlightPair[] resourcePairs =
+        FlightUtils.getRequired(
+            context.getWorkingMap(),
+            RESOURCE_DELETE_FLIGHT_PAIR_LIST,
+            ResourceDeleteFlightPair[].class);
+
+    Stairway stairway = context.getStairway();
+    // Loop through the resources, deleting them one at a time
+    for (ResourceDeleteFlightPair pair : resourcePairs) {
+      try {
+        FlightState flightState = stairway.getFlightState(pair.flightId());
+        FlightStatus flightStatus = flightState.getFlightStatus();
+
+        // If we already finished this flight, we continue through the list of resources to delete
+        if (flightStatus == SUCCESS) {
+          continue;
+        }
+
+        // If the resource delete flight completed with an error. We complete this delete
+        // flight with a different error, reporting the failure
+        if (flightStatus == ERROR || flightStatus == FATAL) {
+          handleResourceFlightFailure(flightState, pair.resourceId());
+        }
+
+        // The flight exists and is still running in Stairway. We wait...
+        waitForFlight(pair, stairway);
+      } catch (FlightNotFoundException e) {
+        launchFlightAndWait(pair, stairway);
+      }
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void handleResourceFlightFailure(FlightState flightState, UUID resourceId) {
+    String message;
+    Throwable throwable;
+    Optional<Exception> flightException = flightState.getException();
+    if (flightException.isPresent()) {
+      throwable = flightException.get();
+      message = throwable.getMessage();
+    } else {
+      throwable = null;
+      message = "flight failed with no error message";
+    }
+    throw new CloudContextDeleteException(
+        String.format(
+            "Cloud context delete failed because resource %s could not be deleted. Cause was: %s",
```

Comment:
Given that this is a message we'll end up returning to callers via `JobReport` error, it might be helpful to also include resource name alongside the ID, since users will actually recognize those

Reply:
Good point. It does require doing a database op on the error path, but I wrapped that in a try/catch so we don't have nested failures flying around.